api-server/scanner-lib/src/blockchain_state/mod.rs

```diff
@@ -154,7 +138,322 @@ impl<S: ApiServerStorage + Send + Sync> LocalBlockchainState for BlockchainState
+    let inputs_utxos = collect_inputs_utxos(db_tx, tx.inputs()).await?;
+
+    let pools = prefetch_pool_amounts(&inputs_utxos, db_tx).await?;
+
+    let pledge_getter = |pool_id: PoolId| Ok(pools.get(&pool_id).cloned());
+    // only used for checks for attempted to print money but we don't need to check that here
+    let delegation_balance_getter = |_delegation_id: DelegationId| Ok(Some(Amount::MAX));
+    // TODO: will be removed as not needed for v0 tokens
+    let issuance_token_id_getter = |_tx_id| Ok(Some(TokenId::new(H256::zero())));
+
+    let inputs_accumulator = ConstrainedValueAccumulator::from_inputs(
+        chain_config,
+        block_height,
+        pledge_getter,
+        delegation_balance_getter,
+        issuance_token_id_getter,
+        tx.inputs(),
+        &inputs_utxos,
+    )
+    .expect("valid block");
+    let outputs_accumulator =
+        ConstrainedValueAccumulator::from_outputs(chain_config, block_height, tx.outputs())
+            .expect("valid block");
+    let consumed_accumulator =
+        inputs_accumulator.satisfy_with(outputs_accumulator).expect("valid block");
+    Ok(consumed_accumulator)
+}
+
+async fn prefetch_pool_amounts<T: ApiServerStorageWrite>(
+    inputs_utxos: &Vec<Option<TxOutput>>,
+    db_tx: &mut T,
+) -> Result<BTreeMap<PoolId, Amount>, ApiServerStorageError> {
+    let mut pools = BTreeMap::new();
+    for output in inputs_utxos {
+        match output {
+            Some(
+                TxOutput::CreateStakePool(pool_id, _) | TxOutput::ProduceBlockFromStake(_, pool_id),
+            ) => {
+                let amount =
+                    db_tx.get_pool_data(*pool_id).await?.expect("should exist").pledge_amount();
+                pools.insert(*pool_id, amount);
+            }
+            Some(
+                TxOutput::Burn(_)
+                | TxOutput::Transfer(_, _)
+                | TxOutput::LockThenTransfer(_, _, _)
+                | TxOutput::DataDeposit(_)
+                | TxOutput::CreateDelegationId(_, _)
+                | TxOutput::DelegateStaking(_, _),
+            ) => {}
+            Some(TxOutput::IssueNft(_, _, _) | TxOutput::IssueFungibleToken(_)) => {
+                //TODO: add when we support tokens
+            }
+            None => {}
+        }
+    }
+    Ok(pools)
+}
+
+async fn collect_inputs_utxos<T: ApiServerStorageWrite>(
+    db_tx: &mut T,
+    inputs: &[TxInput],
+) -> Result<Vec<Option<TxOutput>>, ApiServerStorageError> {
+    let mut outputs = Vec::with_capacity(inputs.len());
+    for input in inputs {
+        let output = match input {
+            TxInput::Utxo(outpoint) => {
+                db_tx.get_utxo(outpoint.clone()).await?.map(|utxo| utxo.into_output())
+            }
+            TxInput::Account(_) | TxInput::AccountCommand(_, _) => None,
+        };
+
+        outputs.push(output);
+    }
+
+    Ok(outputs)
+}
+
+async fn update_tables_from_consensus_data<T: ApiServerStorageWrite>(
```

Comment:
I think that this function is better reuse `tx_verifier::distribute_pos_reward` by reimplementing 2 traits instead of mimicing the logic of distribution and exposing internal functions. It requires some minor changes in the verifier though.